Update README with release, package, certificate information (#53)
jenkins-infra/release: README.adoc

= Jenkins Release
:toc: right

== Introduction

> This repository contains everything needed to release https://github.com/jenkinsci/jenkins[Jenkins core] from Jenkins which includes:

Reviewer: Minor phrasing
```suggestion
This repository contains everything needed to release https://github.com/jenkinsci/jenkins[Jenkins core]:
```

* Jenkinsfiles for Releasing and Packaging.
* Kubernetes POD templates definition for releasing and packaging on Linux and Windows environments.
* Any configuration or utility scripts

The release process implies following Git repository:

> * https://github.com/jenkinsci/jenkins[jenkinsci/jenkins] or any other git repository containing the Jenkins code that needs to be released and or packaged for.

Reviewer: Simplify phrasing
```suggestion
* https://github.com/jenkinsci/jenkins[jenkinsci/jenkins] or any other git repository containing the Jenkins code to be released.
```

> * https://github.com/jenkinsci/packaging[jenkinsci/packaging] contains every tools and configuration used to package Jenkins for OS distribution [Debian,Redhat,Suse,MSI,WAR]

Reviewer: Correct names of operating system distributions
```suggestion
* https://github.com/jenkinsci/packaging[jenkinsci/packaging] contains tools and configuration to package Jenkins for OS distribution [Debian, Red Hat, SUSE, MSI, WAR]
```

* https://github.com/jenkins-infra/chart[jenkins-infra/chart] contains Helm chart used to deploy release.ci.jenkins.io on top of Kubernetes.

> * https://github.com/jenkins-infra/azure[jenkins-infra/azure] contains the terraform code used to provision infrastructures needed by the release environment

Reviewer: ```suggestion
* https://github.com/jenkins-infra/azure[jenkins-infra/azure] contains the terraform code used to provision infrastructure needed by the release environment
```

Remark: It's important to notice that we do not use the maven release plugin to checkout our Jenkins repository neither to push changes as we want to be able to use a different git repository than the one defined in the Jenkins pom.xml

=== Required

- [x] https://github.com/jenkins-infra/openvpn[Jenkins VPN Access]
- [x] Being part of LDAP group 'release-core'
- [x] Access to https://release.ci.jenkins.io[release.ci.jenkins.io]

=== Settings

Those are the most important settings defined in your release profile that will impact your release:

[cols="1,3", options="header"]
.Settings
|===
| Variable | Description

| `JENKINS_GIT_REPOSITORY`
| Defines the Jenkins git repository used by release, to release from and then push generated commits.

| `JENKINS_GIT_BRANCH`
| Defines the Jenkins branch used to checkout and then pushes commits related to the release.

|`MAVEN_REPOSITORY_URL`
| Defines the maven repository URL where it pushes generated artifacts

|`MAVEN_REPOSITORY_NAME`
| Defines the maven repository name where it pushes generated artifacts

|===

=== Steps

Estimated time +- 1h30

> . Open your favorite browser to https://release.ci.jenkins.io

Reviewer: ```suggestion
. Connect to the Jenkins VPN
. Open your favorite browser to https://release.ci.jenkins.io
```

. Trigger the release job on the master branch. https://release.ci.jenkins.io/blue/organizations/jenkins/core-release/branches[Link]

> . Once triggered, it asks you which release line you want to do. It's important to know that the release line matches one of the profile defines https://github.com/jenkins-infra/release/tree/master/profile.d[here], so please carefully review the settings and be sure that is what you are looking for.

Reviewer: ```suggestion
. Once triggered, it asks you which release line you want to do. It's important to know that the release line matches one of the profiles defined https://github.com/jenkins-infra/release/tree/master/profile.d[here], so please carefully review the settings and be sure that is what you are looking for.
```

. At the end of the job, several git commit and maven artifacts will be pushed on their respective locations

=== Validate
To validate that the release went well, except by having a green build, you can double-check that your artifacts have been correctly pushed on your Maven repository located on `$MAVEN_REPOSITORY_URL/$MAVEN_REPOSITORY_NAME/org/jenkins-ci/main/jenkins-war/` and also that running `jarsigner -verify <your generated artifact>` is correct.

=== Stage
Currently, we do not provide an automated way to promote a release but you can still stage a release by specifying a staging git repository and/or a different maven repository. Then you have to promote manually your release.

To promote your git repository, you have to open a pull request from your git repository to https://github.com/jenkinsci/jenkins/compare?expand=1[jenkinsci/jenkins] on the appropriated branch.

To promote your maven artifacts, you have to manually move them to the right location.

== Package

> The packaging process looks after the latest Jenkins version published on a Maven repository and then build and publish artifacts for Debian, Redhat, Suse, Windows and also republish the War file on the package server.

Reviewer: ```suggestion
The packaging process looks after the latest Jenkins version published on a Maven repository and then build and publish artifacts for Debian, Red Hat, SUSE, Windows and also republishes the War file on the package server.
```

- [x] https://github.com/jenkins-infra/openvpn[Jenkins VPN Access]
- [x] Being part of LDAP group 'release-core'
- [x] Access to https://release.ci.jenkins.io[release.ci.jenkins.io]

=== Settings

Those are the most important settings defined in your release profile that will impact the packaging process:

[cols="1,3a", options="header"]
.Settings
|===
| Variable | Description

| JENKINS_VERSION
| Define which version of Jenkins will be downloaded where version accept following value:

* `weekly`, fetches the latest version available on a maven repository.
* `stable`, fetches the latest version that matches pattern X.Y.Z
* `version`, a valid version

https://github.com/jenkins-infra/release/blob/master/utils/getJenkinsVersion.py[getJenkinsVersion.py]

| MAVEN_REPOSITORY_URL
| Defines the maven repository URL used to download the jenkins.war

| MAVEN_REPOSITORY_NAME
| Defines the maven repository name used to download the jenkins.war

| PKGSERVER
| Defines where the different packages will be published

| RELEASELINE
| Define the release line used by packaging scripts in https://github.com/jenkinsci/packaging[jenkinsci/packaging]

|===

=== Steps

Estimated time +- 30min

> . Open your favorite browser to https://release.ci.jenkins.io[release.ci.jenkins.io]

Reviewer: ```suggestion
. Connect to the Jenkins VPN
. Open your favorite browser to https://release.ci.jenkins.io[release.ci.jenkins.io]
```

. Trigger the packaging job on the master branch. https://release.ci.jenkins.io/blue/organizations/jenkins/core-package/branches[Link]
. Once triggered, it asks you which release line you want to package for. The release line matches one of the profile defines in https://github.com/jenkins-infra/release/tree/master/profile.d[profile.d], so please carefully review those settings in order to validate that's what you are looking for.

Once the job is done, every package will be published and then mirror synced

=== Validate
Ensure that packages are correctly published on pkg.jenkins.io and correctly signed.

=== Stage
Staging packages is not supported.

== Certificate

> The Jenkins project uses a Digicert account provided by CDF to request code signing certificate.

Reviewer: ```suggestion
The Jenkins project uses a Digicert account provided by the Continuous Delivery Foundation to request code signing certificates.
```
